### PR TITLE
fix: structure api error messages

### PR DESCRIPTION
**Pull request by Geeker1**
## Description
Structuring error messages from api to users using django-rest custom exception handler.

## Related Issue
[#214]('https://github.com/OpenUpSA/wazimap-ng/issues/212')
## How to test it locally
```docker-compose run --rm -e DJANGO_CONFIGURATION=Test web pytest /app/tests/test_utils.py /app/tests/test_error_handling```

## Changelog
Added test for custom exception handler function
Replaced Custom Renderer with Exception Handler

## Checklist

- [ ]  🚀 is the code ready to be merged and go live?
- [x]  🛠 does it work (build) locally

### Pull Request

- [ ]  📰 good title
- [ ]  📝good description
- [ ]  🔖 issue linked
- [ ]  📖 changelog filled out

### Commits

- [x]  commits are clean
- [x]  commit messages are clean

### Code Quality

- [x]  🚧 no commented out code
- [ ]  🖨 no unnecessary logging
- [ ]  🎱 no magic numbers
- [ ]  black was run locally (as part of the pre-commit hook)

### Testing

- [x]  ✅ added (appropriate) unit tests
- [ ]  💢 edge cases in tests were considered
- [x]  ✅ ran tests locally & are passing

### CODE REVIEW

File: tests/test_response.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+
+@pytest.mark.django_db
+class TestErrorResponseData(APITestCase):
+
+    def test_error_handler(self):
+        factory = RequestFactory()
+        request = factory.get('/api/v1/jkhj')
+
+        response = error_handler(request)
```
Comment (milafrerichs) [inline tests/test_response.py:17]: you're headed in the wrong direction here. 
you want to test the result of the exception raised by DRF and django and not call your own `error_handler`

**Review by milafrerichs — CHANGES_REQUESTED** [review]
good start, but some wrong turns that can be fixed. 

you should add unit tests to `test_utils.py` and rename the `test_response` to `test_error_handling` or so. 

here are some resources for testing Django with DjangoTest plus and DRF:

https://www.django-rest-framework.org/api-guide/testing/
https://www.revsys.com/tidbits/django-test-plus/

File: wazimap_ng/utils.py [hunk 2]
```diff
@@ -615,3 +618,24 @@ def clean_columns(file):
...
+
+def error_handler(request, **kwargs):
+
+    if request.path.startswith('/api/v1'):
+
+        message = 'Server Error occurred.'
+        error_type = 'server_error'
+        code = 500
+
+        data = dict(
```
[the remark this answers is not in the document]
Comment (milafrerichs) [reply wazimap_ng/utils.py:631]: none, it's just a style issue. we/I prefer consistency in code styles.

File: wazimap_ng/utils.py [hunk 3]
```diff
@@ -615,3 +618,24 @@ def clean_columns(file):
     new_columns = df.columns.str.lower().str.strip()
     file.seek(0)
     return old_columns, new_columns
+
+
+def error_handler(request, **kwargs):
+
+    if request.path.startswith('/api/v1'):
```
[the remark this answers is not in the document]
Comment (Geeker1) [reply wazimap_ng/utils.py:625]: Since this fix is to structure all error messages in json, I'll remove the if statement and render all 500 error responses in json.

[the remark this answers is not in the document]
Comment (milafrerichs) [reply wazimap_ng/utils.py]: no, the issue was that all API requests should respond with JSON. 
Not all requests. Otherwise the Admin will show JSON errors and the DRF preview potentially as well.

That's why we need the check for content type here.

File: tests/test_utils.py [hunk 4]
```diff
@@ -65,4 +74,22 @@ def generate_file(data, header, encoding="utf8"):
...
+        fake_request = self.get('/api/fake-request')
+
+        response = error_handler(fake_request)
+        content_type = response['Content-Type']
+
+        assert response.status_code == status.HTTP_500_INTERNAL_SERVER_ERROR
+        assert content_type == 'application/json'
+
+    def test_custom_exception(self):
+        response = self.get('/api/v1/datasets/3452/')
```
Comment (milafrerichs) [inline tests/test_utils.py:92]: this is the integration test and should go to `test_error_handler`

File: tests/test_error_handling.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+from rest_framework import status
+
+
+@pytest.mark.django_db
+class TestErrorResponseData(APITestCase):
+
+    def test_error_json(self):
+        request = self.get('/foo/bar')
+
+        response = error_handler(request)
```
Comment (milafrerichs) [inline tests/test_error_handling.py:15]: this is a unit test since it's calling the `error_handler` method and should go in `test_utils.py`

File: tests/test_error_handling.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            'error': {
+                "message": {
+                    "detail": 'Not found.'
+                },
+                'type': 'not_found',
+                'code': status.HTTP_404_NOT_FOUND
+            }
+        }
+
+        response = self.get('dataset-detail', pk=234)
```
Comment (milafrerichs) [inline tests/test_error_handling.py:35]: this is the correct test for the 404. 
thanks. we need a similar one for the 500

File: wazimap_ng/utils.py [hunk 7]
```diff
@@ -615,3 +620,41 @@ def clean_columns(file):
...
+            'code': code
+        }
+    }
+
+    return JsonResponse(
+        data, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+
+def custom_exception_handler(exc, context):
+    from rest_framework.views import exception_handler
```
Comment (milafrerichs) [inline wazimap_ng/utils.py:644]: why did you put the import here and not at the top with the other imports?

File: tests/test_error_handling.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+        data = json.loads(response.content)
```
Comment (milafrerichs) [inline tests/test_error_handling.py:37]: take a look at the docs from DRF: https://www.django-rest-framework.org/api-guide/testing/#checking-the-response-data

you can use `.data` instead of `.response`

**Review by milafrerichs — CHANGES_REQUESTED** [review]

File: tests/test_error_handling.py (hunk 6 above)
Comment (Geeker1) [reply tests/test_error_handling.py:35]: I might have to do a LiveServer test for 500 errors, during testing, django raises the exception inplace of calling the handler500 function which returns the json response. I wrote the unit test for the handler500 function which is the `test_error_handler_func`, should write the live server test case for 500 error ?

File: wazimap_ng/utils.py (hunk 7 above)
Comment (Geeker1) [reply wazimap_ng/utils.py:644]: It results in a circular import error.

File: wazimap_ng/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Geeker1) [reply wazimap_ng/utils.py]: Fixed. During test , I had to assign content_type to underlying request object so function could be tested without errors.

**Review by goyal1092 — CHANGES_REQUESTED** [review]
@Geeker1 I am not able to see error message on custom api functions like all_detail api end point if I pass invalid porfile id

the solution will work only DRF views but we want to implement it all api endpoints including cutsom api_views

**Comment by adieyal** [comment]
What is the status with this PR - I see it's still in draft - is it still being worked on?

**Comment by Geeker1** [comment]
> What is the status with this PR - I see it's still in draft - is it still being worked on?

Yes, it is. Will soon push updates
